tools/envoy_headersplit/headersplit.py

```diff
@@ -0,0 +1,367 @@
+# !/usr/bin/env python3
+# Lint as: python3
+"""
+This python script can dividing monolithic mock headers
+into different mock classes. We need to remove the
+over-included head files in generated class codes and
+resolve dependencies in the corresponding Bazel files
+manually.
+"""
+
+from __future__ import print_function
+
+import argparse
+from typing import Type, List, Tuple, Dict
+import clang.cindex
+from clang.cindex import TranslationUnit, Index, CursorKind, Cursor
+
+clang.cindex.Config.set_library_path("/opt/llvm/lib")
+
+
+def to_filename(classname: str) -> str:
+  """
+    maps mock class name (in C++ codes) to filenames under the Envoy naming convention.
+    e.g. map "MockAdminStream" to "admin_stream"
+
+    Args:
+        classname: mock class name from source
+
+    Returns:
+        corresponding file name
+    """
+  filename = classname.replace('Mock', '', 1)  # Remove only first "Mock"
+  ret = ""
+  for index, val in enumerate(filename):
+    if val.isupper() and index > 0:
+      ret += '_'
+    ret += val
+  return ret.lower()
+
+
+def get_headers(translation_unit: Type[TranslationUnit]) -> str:
+  """
+    extracts all head includes statements from the target code file (translation_unit)
+
+    for instance:
+        foo.h:
+        #include "a.h"
+        #include "b.h"
+
+        int foo(){
+
+        }
+    this function should return
+    '#include "a.h"\n#include "b.h"'
+
+    Args:
+        translation_unit: parsing result of target source code by libclang
+
+    Returns:
+        A string, contains all includes statements from the source code.
+
+    """
+
+  # clang lib provides API like tranlation_unit.get_inludes()
+  # But we can't use it since it requires presence of the included files to return the full list
+
+  cursor = translation_unit.cursor
+  for descendant in cursor.walk_preorder():
+    if descendant.location.file is not None and descendant.location.file.name == cursor.displayname:
+      filename = descendant.location.file.name
+      with open(filename, 'r') as source_file:
+        contents = source_file.read()
+      headers = contents[:descendant.extent.start.offset]
+      return headers
+
+  return ""
+
+
+def class_definitions(cursor: Cursor) -> List[Cursor]:
+  """
+    extracts all class definitions in the file pointed by cursor. (typical mocks.h)
+
+    Args:
+        cursor: cursor of parsing result of target souce code by libclang
+
+    Returns:
+        a list of cursor, each pointing to a class definition.
+
+    """
+  class_cursors = []
+  for i in cursor.walk_preorder():
+    if i.location.file is None:
+      continue
+    if i.location.file.name != cursor.displayname:
```

Comment:
What does this check do? This section might benefit from some additional comments.

Reply:
i have added comments explaining those checks.